design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+     maintained to the platform as a `workload` or `trait`. Therefore, our solution needs to
+     accommodate pre-defined CRDs that OAM users bring.
+
+## Proposal
+The overall idea is for the applicationConfiguration controller to annotate critical information
+in the workload and trait CR it emits. In addition, we will provide a helper library so that
+trait controller developers can locate the resources they need with a simple function call. 
+Here is the list of changes that we propose.
+1. Add a `childResources` field in the  WorkloadDefinition. 
+Currently, a workloadDefinition is nothing but a shim of a real workload CRD. We propose to add
+an **optional** field called `childResources` to the schema of the workloadDefinition. We encourage
+workload owners to fill in this field when they register their controllers to the OAM system. 
+This is the way for them to declare the types of the Kubernetes resources their workload
+controller actually generates. In our example, the workload definition can claim to generate
+deployment and service child resources. 
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: WorkloadDefinition
+    metadata:
+       name: mydb.standard.oam.dev
+    spec:
+      definitionRef:
+         name: mydb.standard.oam.dev
+      childResources:
+         - groupVersion: apps/v1
+           kind: deployment
+         - groupVersion: core/v1
+           kind: service      
+    ```
+
+2. ApplicationConfig controller add annotations containing workload child resource information to
+the emitted `workload` CR. For each `workload` CR, the applicationConfig controller can fetch the
+corresponding `workloadDefinition` CR following an 
+[OAM convention](https://github.com/oam-dev/spec/blob/master/3.workload.md#definitionref). The
+conventions says that the name of the `workloadDefinition` CR is the name of the `workload` CRD
+it refers to. For example, the name of the `workloadDefinition` CR that refers to a
+`containerizedworkloads.core.oam.dev` CRD is exactly `containerizedworkloads.core.oam.dev` as
+well. In our example, the applicationConfig controller will emit the following `workload` CR with all
+ its child resources GVK in its metadata.
+    ```yaml
+    apiVersion: standard.oam.dev/v1alpha2
+    kind: Mydb
+    metadata:
+      name: mydb-example
+      annotations:
+        core.oam.dev/childResources: "{ apps/v1.deployment, core/v1.service}"
+    ```
+
+3. ApplicationConfig controller add annotations containing the workload information to the emitted
+`trait` CR. The application controller actually generates the workload CR so it has all the
+information. In our example, the applicationConfig controller will emit the following `trait` CR with
+the GVK and name of the workload the `ManualScalerTrait` is applied to.
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ManualScalerTrait
+    metadata:
+      annotations:
```

Comment:
It seems like we're taking the contemporary `spec.workloadRef` of OAM-aware traits like `ManualScalerTrait` and moving them into annotations. I believe the idea is that we want to allow existing Kubernetes custom resources (that are blissfully unaware of OAM) to act as traits, similar to how existing OAM-unaware custom resources can act as OAM workloads.

How does this mesh with the below proposal that trait developers will use an OAM-specific library to lookup the workload to which a trait applies? It sounds like this data will only be used by developers who are intending to write an OAM trait. Presumably an existing OAM-unaware Kubernetes CR would not be aware of these annotations.

My suspicion is that if it's possible for an existing Kubernetes CR to act as an OAM trait, then that CR will already have some way of indicating what resource it operates on, for example a label selector in its spec.